File: test/src/scala/src/main/scala/com/baidu/inf/Bayes.scala
package com.baidu.inf

import org.apache.spark.mllib.classification.NaiveBayes
import org.apache.spark.mllib.linalg.Vectors
import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.{SparkContext,SparkConf}

object Bayes {
  def main(args: Array[String]) {
    if (args.length < 2) {
      println("usage: spark-submit com.baidu.inf.WordCount <input> <output>")
      System.exit(1)
    }
    val conf =new SparkConf()
    val sc =new SparkContext(conf)
    
    var input_path = args(0)
    var output_path = args(1)

    //读入数据
    val data = sc.textFile(input_path)
    val parsedData =data.map { line =>
      val parts =line.split(',')
      LabeledPoint(parts(0).toDouble,Vectors.dense(parts(1).split(' ').map(_.toDouble)))
    }
    // 把数据的60%作为训练集，40%作为测试集.
    val splits = parsedData.randomSplit(Array(0.6,0.4),seed = 11L)
    val training =splits(0)
    val test =splits(1)

    //获得训练模型,第一个参数为数据，第二个参数为平滑参数，默认为1，可改
    val model =NaiveBayes.train(training,lambda = 1.0)

    //对模型进行准确度分析
    val predictionAndLabel= test.map(p => (model.predict(p.features),p.label))
    val accuracy =1.0 *predictionAndLabel.filter(x => x._1 == x._2).count() / test.count()

    println("accuracy-->"+accuracy)
    val result = model.predict(Vectors.dense(0.0,2.0,0.0,1.0))
    println("Predictionof (0.0, 2.0, 0.0, 1.0):"+ result)

    // Save and load model
    //model.save(sc, "target/tmp/myNaiveBayesModel")
    //val sameModel = NaiveBayesModel.load(sc, "target/tmp/myNaiveBayesModel")
    val pv = (accuracy,result)
    val out = sc.parallelize( List(pv) )
    
    out.saveAsTextFile(output_path)
  }
}

File: test/src/pyspark/bayes.py
# Copyright (c) 2017 Baidu, Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import print_function

from pyspark import SparkContext
# $example on$
from pyspark.mllib.classification import NaiveBayes, NaiveBayesModel
from pyspark.mllib.linalg import Vectors
from pyspark.mllib.regression import LabeledPoint


def parseLine(line):
    parts = line.split(',')
    label = float(parts[0])
    features = Vectors.dense([float(x) for x in parts[1].split(' ')])
    return LabeledPoint(label, features)

if __name__ == "__main__":
    input_path = "hdfs:///bayes_input/"
    output_path = "hdfs:///bayes_output/"

    sc = SparkContext(appName="PythonNaiveBayesExample")

    # $example on$
    data = sc.textFile(input_path).map(parseLine)

    # Split data aproximately into training (60%) and test (40%)
    training, test = data.randomSplit([0.6, 0.4], seed=0)

    # Train a naive Bayes model.
    model = NaiveBayes.train(training, 1.0)

    # Make prediction and test accuracy.
    predictionAndLabel = test.map(lambda p: (model.predict(p.features), p.label))
    accuracy = 1.0 * predictionAndLabel.filter(lambda (x, v): x == v).count() / test.count()

    result = model.predict(Vectors.dense(0.0,2.0,0.0,1.0))

    print("accuracy-->" + str(accuracy))
    print("Predictionof (0.0, 2.0, 0.0, 1.0):"+ str(result))

    # Save and load model
    #model.save(sc, "target/tmp/myNaiveBayesModel")
    #val sameModel = NaiveBayesModel.load(sc, "target/tmp/myNaiveBayesModel")
    pv = (accuracy,result)
    out = sc.parallelize( [pv] )
    out.saveAsTextFile(output_path)
    # Save and load model
    #model.save(sc, "/home/miaodongdong/data/bayes/myNaiveBayesModel")
    #sameModel = NaiveBayesModel.load(sc, "/home/miaodongdong/data/bayes/myNaiveBayesModel")
